test/integration/start_tls_integration_test.cc

```diff
@@ -0,0 +1,533 @@
+public:
+  // Network::ReadFilter
+  Network::FilterStatus onData(Buffer::Instance& data, bool end_stream) override;
+  Network::FilterStatus onNewConnection() override;
+  void initializeReadFilterCallbacks(Network::ReadFilterCallbacks& callbacks) override {
+    read_callbacks_ = &callbacks;
+  }
+
+  // Network::WriteFilter
+  Network::FilterStatus onWrite(Buffer::Instance& data, bool end_stream) override;
+  void initializeWriteFilterCallbacks(Network::WriteFilterCallbacks& callbacks) override {
+    write_callbacks_ = &callbacks;
+  }
+
+private:
+  Network::ReadFilterCallbacks* read_callbacks_{};
+  Network::WriteFilterCallbacks* write_callbacks_{};
+};
+
+// Network::ReadFilter
+Network::FilterStatus StartTlsSwitchFilter::onNewConnection() { return Network::FilterStatus::Continue; }
+
+Network::FilterStatus StartTlsSwitchFilter::onData(Buffer::Instance&, bool) {
+  printf("Data Received onData!\n");
+//  data.drain(data.length());
+  printf("%s\n", read_callbacks_->connection().transportProtocol().c_str());
+  if (read_callbacks_->connection().transportProtocol() ==
+      Extensions::TransportSockets::TransportProtocolNames::get().StartTls) {
+    // We run on top of the `STARTTLS` socket and can ask it to convert to SSL.
+  printf("Switching to TLS!\n");
+    read_callbacks_->connection().startSecureTransport();
+  }
+#if 0
+#endif
+
+  return Network::FilterStatus::Continue;
+}
+
+// Network::WriteFilter
+Network::FilterStatus StartTlsSwitchFilter::onWrite(Buffer::Instance&, bool) {
+  printf("Data Received onWrite!\n");
+  //data.drain(data.length());
+  return Network::FilterStatus::Continue;
+}
+
+/**
+ * Config factory for StartTlsSwitchFilter.
+ */
+class StartTlsSwitchFilterConfigFactory : public Extensions::NetworkFilters::Common::FactoryBase<
+                                         test::integration::start_tls::StartTlsFilterConfig> {
+public:
+  explicit StartTlsSwitchFilterConfigFactory(const std::string& name) :  FactoryBase(name) {}
+
+  Network::FilterFactoryCb createFilterFactoryFromProtoTyped(
+      const test::integration::start_tls::StartTlsFilterConfig& ,
+      Server::Configuration::FactoryContext&) override {
+    return [](Network::FilterManager& filter_manager) -> void {
+      filter_manager.addFilter(std::make_shared<StartTlsSwitchFilter>());
+    };
+  }
+
+/*
+  // NamedNetworkFilterConfigFactory
+  Network::FilterFactoryCb
+  createFilterFactoryFromProto(const Protobuf::Message&,
+                               Server::Configuration::FactoryContext&) override {
+    return [](Network::FilterManager& filter_manager) -> void {
+      filter_manager.addFilter(std::make_shared<StartTlsSwitchFilter>());
+    };
+  }
+
+  ProtobufTypes::MessagePtr createEmptyConfigProto() override {
+    // Using Struct instead of a custom per-filter empty config proto
+    // This is only allowed in tests.
+    return ProtobufTypes::MessagePtr{new Envoy::ProtobufWkt::Struct()};
+  }
+*/
```

Comment:
Do we need this?

Reply:
The same as above: I think you are reviewing NOT the last version.

Reply:
Oops... sorry by the noise!!!